lib/internal/fs/glob.js

```diff
@@ -0,0 +1,202 @@
+                             (pattern) => ({ __proto__: null, pattern, index: 0, path: '.', followSymlinks: true }));
+  });
+  const cache = new Cache(matchers);
+
+  while (queue.length > 0) {
+    const { pattern, index: currentIndex, path, followSymlinks } = ArrayPrototypePop(queue);
+    if (cache.seen(pattern, currentIndex, path)) {
+      continue;
+    }
+    cache.add(pattern, currentIndex, path);
+
+    const currentPattern = pattern[currentIndex];
+    const index = currentIndex + 1;
+    const isLast = pattern.length === index || (pattern.length === index + 1 && pattern[index] === '');
+
+    if (currentPattern === '') {
+      // Absolute path
+      ArrayPrototypePush(queue, { __proto__: null, pattern, index, path: '/', followSymlinks });
+      continue;
+    }
+
+    if (typeof currentPattern === 'string') {
+      const entryPath = join(path, currentPattern);
+      if (isLast && cache.stats(resolve(root, entryPath))) {
+        // last path
+        results.add(entryPath);
+      } else if (!isLast) {
+        // Keep traversing, we only check file existence for the last path
+        ArrayPrototypePush(queue, { __proto__: null, pattern, index, path: entryPath, followSymlinks });
+      }
+      continue;
+    }
+
+    const fullpath = resolve(root, path);
+    const stat = cache.stats(fullpath);
+    const isDirectory = stat?.isDirectory() || (followSymlinks !== false && stat?.isSymbolicLink());
+
+    if (isDirectory && isRegExp(currentPattern)) {
+      const entries = cache.readdir(fullpath);
+      for (const entry of entries) {
+        const entryPath = join(path, entry.name);
+        if (cache.seen(pattern, index, entryPath)) {
+          continue;
+        }
+        const matches = testPattern(currentPattern, entry.name);
+        if (matches && isLast) {
+          results.add(entryPath);
+        } else if (matches) {
+          ArrayPrototypePush(queue, { __proto__: null, pattern, index, path: entryPath, followSymlinks });
+        }
+      }
+    }
+
+    if (currentPattern === GLOBSTAR && isDirectory) {
+      const entries = cache.readdir(fullpath);
+      for (const entry of entries) {
+        if (entry.name[0] === '.' || (exclude && exclude(entry.name))) {
+          continue;
+        }
+        const entryPath = join(path, entry.name);
+        if (cache.seen(pattern, index, entryPath)) {
+          continue;
+        }
+        const isSymbolicLink = entry.isSymbolicLink();
+        const isDirectory = entry.isDirectory();
+        if (isDirectory) {
+          // Push child directory to queue at same pattern index
+          ArrayPrototypePush(queue, {
+            __proto__: null, pattern, index: currentIndex, path: entryPath, followSymlinks: !isSymbolicLink,
+          });
+        }
+
+        if (pattern.length === index || (isSymbolicLink && pattern.length === index + 1 && pattern[index] === '')) {
+          results.add(entryPath);
+        } else if (pattern[index] === '..') {
```

Comment:
Is this testing for stuff like `a/**/../x`?  If so, it should be filtering for directories.

Eg, `a/**/../x` will match only if the `**` can match a directory.  If `a` doesn't contain any directories, it won't find a match, even if `a/x` exists.  This is made somewhat more efficient where it can if you pass `optimizationLevel: 2` to the Minimatch constructor; any `..` that _can_ be pre-emptively stripped out will be, and any `..` that remain must be processed by returning the processing to the parent directory.  This handles the fact that `root/a/**/../x/y` is equivalent to `{root/x/y,root/a/**/x/y}`, but `root/a/**/../x` is not equivalent to `{root/x,root/a/**/x}`.  The extra `y` portion ensures that `x` must be a directory, so we know that there'll be at least one directory below `a`.

Reply:
Probably better in this instance to treat `..` as "just another string path portion", tack it on and include it in the matches if it stats successfully.